cli/src/audiostream.c

```diff
@@ -0,0 +1,155 @@
+#define ARG_KEY_MORNING 'm'
+
+static struct argp_option options[] = {
+	{ "host", ARG_KEY_HOST, "Host", 0, "Host to request audiostream from", 0 },
+	{ "registkey", ARG_KEY_REGISTKEY, "RegistKey", 0, "PS4 registration key", 0 },
+	{ "morning", ARG_KEY_MORNING, "Morning", 0, "PS4 morning encoded in base64", 0 },
+	{ 0 }
+};
+
+typedef struct arguments
+{
+	const char *host;
+	const char *registkey;
+	const char *morning;
+} Arguments;
+
+static int parse_opt(int key, char *arg, struct argp_state *state)
+{
+	Arguments *arguments = state->input;
+
+	switch(key)
+	{
+		case ARG_KEY_HOST:
+			arguments->host = arg;
+			break;
+		case ARG_KEY_REGISTKEY:
+			arguments->registkey = arg;
+			break;
+		case ARG_KEY_MORNING:
+			arguments->morning = arg;
+			break;
+		case ARGP_KEY_ARG:
+			argp_usage(state);
+			break;
+		default:
+			return ARGP_ERR_UNKNOWN;
+	}
+
+	return 0;
+}
+
+static struct argp argp = { options, parse_opt, 0, doc, 0, 0, 0 };
+
+static void AudioFrameCb(int16_t *buf, size_t samples_count, void *user)
+{
+	fprintf(stderr, "%s", (char *)buf);
+}
+
+CHIAKI_EXPORT int chiaki_cli_cmd_audiostream(ChiakiLog *log, int argc, char *argv[])
+{
+	Arguments arguments = { 0 };
+	error_t argp_r = argp_parse(&argp, argc, argv, ARGP_IN_ORDER, NULL, &arguments);
+	if(argp_r != 0)
+		return 1;
+
+	if(!arguments.host)
+	{
+		fprintf(stderr, "No host specified, see --help.\n");
+		return 1;
+	}
+	if(!arguments.registkey)
+	{
+		fprintf(stderr, "No registration key specified, see --help.\n");
+		return 1;
+	}
+	if(!arguments.morning)
+	{
+		fprintf(stderr, "No morning specified, see --help.\n");
+		return 1;
+	}
+	if(strlen(arguments.registkey) > 8)
+	{
+		fprintf(stderr, "Given registkey is too long.\n");
+		return 1;
+	}
+
+	ChiakiConnectInfo chiaki_connect_info;
+	size_t morning_size = sizeof(chiaki_connect_info.morning);
+	uint8_t morning[morning_size];
+	chiaki_base64_decode(arguments.morning, strlen(arguments.morning), morning, &morning_size);
+	memcpy(chiaki_connect_info.morning, morning, morning_size);
+	chiaki_connect_info.host = arguments.host;
+	memcpy(chiaki_connect_info.regist_key, arguments.registkey, sizeof(chiaki_connect_info.regist_key));
+
+	ChiakiConnectVideoProfile video_profile;
+	// From what it seems like, PS4 does not allow for audio-only stream.
+	// So it's necessary to also request for video stream. This needs to
+	// be confirmed though.
+	video_profile.width = 640;
+	video_profile.height = 360;
+	video_profile.max_fps = 30;
+	video_profile.bitrate = 128;
+	chiaki_connect_info.video_profile = video_profile;
+
+	ChiakiSession session;
+	ChiakiErrorCode err;
+	err = chiaki_session_init(&session, &chiaki_connect_info, NULL);
+	if(err != CHIAKI_ERR_SUCCESS)
+		return 2;
+
+	ChiakiOpusDecoder opus_decoder;
+	chiaki_opus_decoder_init(&opus_decoder, NULL);
+	chiaki_opus_decoder_set_cb(&opus_decoder, NULL, AudioFrameCb, NULL);
+
+	ChiakiAudioSink audio_sink;
+	chiaki_opus_decoder_get_sink(&opus_decoder, &audio_sink);
```

Comment:
Instead of using `ChiakiOpusDecoder`, which would decode the opus itself, you will want to implement the `ChiakiAudioSink` interface yourself and then output the raw opus data in `ChiakiAudioSink.frame_cb`. You might have to also craft a header for ogg or something in `ChiakiAudioSink.header_cb`, so the reading program will be able to determine how to deal with the data.

Reply:
Alright, thanks! I'll give this a shot.

Reply:
I can't seem to properly craft the audio stream header. It seems like we are crafting the header for audio stream in the android build in:
https://github.com/thestr4ng3r/chiaki/blob/d4b4681a0f6a527620ace84ba24ffa1df08ce258/android/app/src/main/cpp/audio-decoder.c#L149-L175

So I mostly copy-pasted the same thing here but I still can't seem to get the audio working with pipes.